bmap-writer.cpp

```diff
@@ -185,66 +206,106 @@ int BmapWriteImage(const std::string &imageFile, const bmap_t &bmap, const std::
         }
 
         for (const auto &range : bmap.ranges) {
-            size_t startBlock, endBlock;
-            if (sscanf(range.range.c_str(), "%zu-%zu", &startBlock, &endBlock) == 1) {
-                endBlock = startBlock;  // Handle single block range
+            const size_t outStart = range.startBlock * bmap.blockSize;
+            const size_t outEnd = ((range.endBlock + 1) * bmap.blockSize);
+            const size_t rangeSize = (range.endBlock - range.startBlock + 1) * bmap.blockSize;
+            const off_t writeOffset = static_cast<off_t>(range.startBlock * bmap.blockSize);
+            size_t writtenSize = 0;
+            bool endOfFile = false;
+
+            if (checksumInit(&checksum) < 0) {
+                throw std::string("Failed to init checksum engine");
             }
-            //std::cout << "Processing Range: startBlock=" << startBlock << ", endBlock=" << endBlock << std::endl;
-
-            size_t bufferSize = (endBlock - startBlock + 1) * bmap.blockSize;
-            std::vector<char> buffer(bufferSize);
-            size_t outBytes = 0;
-
-            const size_t outStart = startBlock * bmap.blockSize;
-            const size_t outEnd = ((endBlock + 1) * bmap.blockSize);
-
-            while (outBytes < bufferSize) {
-                ssize_t readData = archive_read_data(a, buffer.data() + outBytes, bufferSize - outBytes);
-
-                // If no more data is available in the input buffer and the input file has been
-                // read completely, stop this decompression loop
-                if (readData <= 0)
-                    break;
-
-                size_t chunkSize = static_cast<size_t>(readData);
-
-                if (decHead >= outStart && (decHead + chunkSize) <= outEnd) {
-                    // Case 1: all decoded data can be used
-                    outBytes += chunkSize;
-                } else if (decHead < outStart && (decHead + chunkSize) <= outStart) {
-                    // Case 2: all decoded data shall be discarded
-                } else if (decHead < outStart && (decHead + chunkSize) > outStart) {
-                    // Case 3: only the last portion of the decoded data can be used
-                    std::move(buffer.begin() + static_cast<long int>(outStart - decHead),
-                              buffer.begin() + static_cast<long int>(chunkSize),
-                              buffer.begin());
-                    size_t validData = chunkSize - (outStart - decHead);
-                    outBytes += validData;
+
+            //std::cout << "Processing Range: startBlock=" << range.startBlock << ", endBlock=" << range.endBlock << std::endl;
+
+            while ((writtenSize < rangeSize) && !endOfFile) {
```

Comment:
can't we have at each range loop a calculation of available free memory and adjust it accordingly ? thus we can avoid at all the parameters _maxbuffersize_ ?

Reply:
Done in updated MR.